### PR TITLE
[Federation] Kubefed init verifies if control plane pods are up before returning success

### PR DESCRIPTION
**Pull request by irfanurrehman**
This PR updates the functionality as needed in issue https://github.com/kubernetes/kubernetes/issues/37841.

cc @kubernetes/sig-cluster-federation @nikhiljindal @madhusudancs @shashidharatd

### CODE REVIEW

File: federation/pkg/kubefed/init/init.go [hunk 1]
```diff
@@ -230,6 +230,23 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 	}
 
 	if !dryRun {
+		fmt.Fprintf(cmdOut, "Waiting for control plane to come up")
+		for !podRunning(hostClientset, serverName, initFlags.FederationSystemNamespace) {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Please use https://godoc.org/k8s.io/kubernetes/pkg/util/wait#Poll or https://godoc.org/k8s.io/kubernetes/pkg/util/wait#PollImmediate instead.

**Review by madhusudancs — COMMENTED** [review]
This might be useful, but isn't sufficient. What we should really do is component health checks, in the same vein as `ComponentStatus`es reported by Kubernetes components. We are not there yet, so what we should do is pod status check you are doing here plus also verify that an API endpoint responds. See here - https://github.com/kubernetes/kubernetes/pull/34484/files#diff-776638e5b09cfdf82b07554f70d6b558R430

**Comment by irfanurrehman** [comment]
@madhusudancs thanks for the comments, have updated accordingly, re-review needed!!

File: federation/pkg/kubefed/init/init.go [hunk 2]
```diff
@@ -229,7 +230,25 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 		return err
 	}
 
+	//to talk to federation ctrl plane
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Let's try to avoid arbitrary acronyms and vowel-omissions in the comments in the future. Comments are there to explain what is happening in the code. Let's not make it difficult to even read them.

File: federation/pkg/kubefed/init/init.go [hunk 3]
```diff
@@ -229,7 +230,25 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
...
+	fedFactory := config.HostFactory(initFlags.Name, initFlags.Kubeconfig)
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Why build it here and not inside the `if`-block, right before the call to `waitSrvHealthy()`? I think it is Ok to just pass the context name and kubeconfig you are passing to `HostFactory()` to `waitSrvHealthy()` instead and build the client object there. 

Also, I am somewhat worried about doing this. `HostFactory()` method was written keeping interactions with host cluster in mind. It wasn't for federation API server. Also, the discovery interface for the command factory builds the objects for Kubernetes APIs and not federation APIs. All of these might work today because of the particular way things are hooked up, but might break down when some of these assumptions are violated.

In order to avoid this problem in the future and also to do the right thing, directly build the federation Clientset in `k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset`. You can then call `fedClientset.Discovery()` to get the federation discovery client.

File: federation/pkg/kubefed/init/init.go [hunk 4]
```diff
@@ -582,6 +601,42 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
 	return clientset.Extensions().Deployments(namespace).Create(dep)
 }
 
+func waitForPods(clientset *client.Clientset, name, nameSpace string) error {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Why not make `waitForPods()` take a slice of names `waitForPods(..., name []string, ...)` and check for both the pods together instead of having to call them separately? Pluralized name for the function is already suggesting that.

Also, convention is to call the variable `namespace` - no capitalization for `S`.

File: federation/pkg/kubefed/init/init.go [hunk 5]
```diff
@@ -582,6 +601,42 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
...
+	err := wait.PollInfinite(podWaitInterval, func() (bool, error) {
+		podList, err := clientset.Core().Pods(nameSpace).List(api.ListOptions{})
+		if err != nil {
+			return false, nil
+		}
+		for _, pod := range podList.Items {
+			if strings.Contains(pod.Name, name) && pod.Status.Phase == "Running" {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: I have been thinking quite a bit about this. I think the right thing to do here is to get the deployment status, then get the pod by using its selector and then check the pod's phase. It's Ok to do that in a follow up PR. But at least do a `strings.HasPrefix()` check instead of a `strings.Contains()` check here.

File: federation/pkg/kubefed/init/init.go [hunk 6]
```diff
@@ -582,6 +601,42 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
...
+	fedDiscoveryClient, err := fedFactory.DiscoveryClient()
+	if err != nil {
+		return err
+	}
+
+	err = wait.PollInfinite(podWaitInterval, func() (bool, error) {
+		body, err := fedDiscoveryClient.RESTClient().Get().AbsPath("/healthz").Do().Raw()
+		if err != nil {
+			return false, nil
+		} else {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Drop the `else` here. It is not required. Logic will still be the same.

File: federation/pkg/kubefed/init/init_test.go [hunk 7]
```diff
@@ -168,7 +168,7 @@ func TestInitFederation(t *testing.T) {
 				want = fmt.Sprintf("Federation control plane runs (dry run)\n")
 			}
 
-			if got := buf.String(); got != want {
+			if got := buf.String(); !strings.Contains(got, want) {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init_test.go]: Why this change?

File: federation/pkg/kubefed/init/init_test.go [hunk 8]
```diff
@@ -694,6 +694,38 @@ func fakeInitHostFactory(federationName, namespaceName, ip, dnsZoneName, image,
...
+		ObjectMeta: v1.ObjectMeta{
+			Name:      svcName,
+			Namespace: namespaceName,
+		},
+		Status: v1.PodStatus{
+			Phase: "Running",
+		},
+	}
+
+	ctrlMgrPod := v1.Pod{
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init_test.go]: `cm` is the convention used for controller manager.

**Review by madhusudancs — CHANGES_REQUESTED** [review]
I have left a few comments. Let me know once you have addressed them.

File: federation/pkg/kubefed/init/init.go (hunk 2 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: sure madhu; will update with mode elaborate comment, and will take care of this in future also.

File: federation/pkg/kubefed/init/init.go (hunk 3 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: > Why build it here and not inside the if-block, right before the call to waitSrvHealthy()? I think it is Ok to just pass the context name and kubeconfig you are passing to HostFactory() to waitSrvHealthy() instead and build the client object there.

Will update accordingly!

Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: > Also, I am somewhat worried about doing this. HostFactory() method was written keeping interactions with host cluster in mind. It wasn't for federation API server. Also, the discovery interface for the command factory builds the objects for Kubernetes APIs and not federation APIs. All of these might work today because of the particular way things are hooked up, but might break down when some of these assumptions are violated.
> 
> In order to avoid this problem in the future and also to do the right thing, directly build the federation Clientset in k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset. You can then call fedClientset.Discovery() to get the federation discovery client.

I have a slight confusion about this. Please help me understand better.
Actually when I implemented it first, I did something similar what are asking for. I did extend the AdminConfig interface to add another method getconfig, which built the config on similar lines as is currently done to get hostfactory, and then got a separate client for fed control plane from that config. I however removed that in favor of simplicity of the current one.
Also on futher checking, the HostFactory() impl gets the new factory from package kubectl/cmd which I guess "should" work with either federation API server or kube API server, for that particular version alike.

I would any ways keep the impl as suggested by you also ready, and update this commit or not, based on your suggestion.

File: federation/pkg/kubefed/init/init.go (hunk 4 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: will update accordingly!!

File: federation/pkg/kubefed/init/init.go (hunk 5 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: will update to use HasPrefix now and update rest of the checking method in a separate PR.

File: federation/pkg/kubefed/init/init.go (hunk 6 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: sure!

File: federation/pkg/kubefed/init/init_test.go (hunk 7 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init_test.go]: Earlier kubefed init will either return a success statement exactly as implemented in printSuccess() or some failure message.
I did find some sense in adding additional messages before it (eg - Waiting for control plane to come up) , else it seemed like hung, if the wait takes long.
I did contemplate adding a "." every second or two to reflect that something is still happening, but did not include it in the final PR somehow.
The check in that case cannot be strictly only on the specific message.
What do u suggest?, should I add that back. or put a strict check on the message on the unit test.

File: federation/pkg/kubefed/init/init_test.go (hunk 8 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init_test.go]: will update!

**Comment by irfanurrehman** [comment]
@madhusudancs , I have updated the PR as per you review comments and suggestions.
As of now, I have added another commit to make it easier for you track changes.
One you are ok and say so, I will squash the review fixes commits into 1.
thanks!

**Comment by madhusudancs** [comment]
@k8s-bot ok to test

File: federation/pkg/kubefed/util/util.go [hunk 9]
```diff
@@ -64,17 +68,30 @@ func (a *adminConfig) PathOptions() *clientcmd.PathOptions {
 	return a.pathOptions
 }
 
-func (a *adminConfig) HostFactory(host, kubeconfigPath string) cmdutil.Factory {
+func (a *adminConfig) FedClientSet(context, kubeconfigPath string) (*fedclient.Clientset, error) {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/util/util.go]: Call this function `FederationClientset`. Expand fed and lower s in set. It is better for readability and Clientset follows the convention elsewhere.

**Review by madhusudancs — COMMENTED** [review]
A few minor comments.

Also, github won't let me reply to a comment on an updated hunk, so replying here:

> I have a slight confusion about this. Please help me understand better.
> Actually when I implemented it first, I did something similar what are asking for. I did extend the AdminConfig interface to add another method getconfig, which built the config on similar lines as is currently done to get hostfactory, and then got a separate client for fed control plane from that config. I however removed that in favor of simplicity of the current one.
> Also on futher checking, the HostFactory() impl gets the new factory from package kubectl/cmd which I guess "should" work with either federation API server or kube API server, for that particular version alike.
>
> I would any ways keep the impl as suggested by you also ready, and update this commit or not, based on your suggestion.

I agree about the simplicity. But the trouble here is, the client and the API types returned by `kubetcl/cmd.Factory` are for Kubernetes. It all works because of the way things are hooked up today. If those assumptions are violated in the future, they might break. There is no promise about stability here. `cmd.Factory` is an internal thing. It can change. So I think we shouldn't risk that.

File: federation/pkg/kubefed/init/init_test.go (hunk 7 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init_test.go]: have updated accordingly, will raise another PR for this change.

File: federation/pkg/kubefed/init/init.go [hunk 10]
```diff
@@ -229,7 +230,20 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 		return err
 	}
 
+	fedPods := []string{}
```
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: done

File: federation/pkg/kubefed/init/init.go [hunk 11]
```diff
@@ -576,6 +590,47 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
 	return clientset.Extensions().Deployments(namespace).Create(dep)
 }
 
+func waitForPods(clientset *client.Clientset, fedPods []string, namespace string) error {
+	for _, fedPod := range fedPods {
```
Comment (irfanurrehman) [inline federation/pkg/kubefed/init/init.go]: updated

File: federation/pkg/kubefed/util/util.go (hunk 9 above)
Comment (irfanurrehman) [inline federation/pkg/kubefed/util/util.go]: updated

**Comment by irfanurrehman** [comment]
@madhusudancs updated as per your suggestions and squashed the review fix commits also. PTAL.

File: federation/pkg/kubefed/init/init.go [hunk 12]
```diff
@@ -576,6 +586,43 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
...
+func waitForPods(clientset *client.Clientset, fedPods []string, namespace string) error {
+	err := wait.PollInfinite(podWaitInterval, func() (bool, error) {
+		podList, err := clientset.Core().Pods(namespace).List(api.ListOptions{})
+		if err != nil {
+			return false, nil
+		}
+		for _, pod := range podList.Items {
+			for _, fedPod := range fedPods {
+				if strings.HasPrefix(pod.Name, fedPod) && pod.Status.Phase == "Running" {
+					return true, nil
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: As I was pointing out in my previous comment, you will need additional logic here. What is happening here is, the poll loop exits due to the return statement as soon as a single `Running` pod is found. However, you need to verify that both the pods are running.

**Review by madhusudancs — COMMENTED** [review]
@irfanurrehman one minor comment, everything else looks good. Please fix that and let's get this merged.

**Comment by irfanurrehman** [comment]
@madhusudancs updated. Also apologies for taking your additional time.

**Comment by madhusudancs** [comment]
@irfanurrehman no worries. Thanks for making these changes and for being patient through these reviews. LGTM!

**Review by madhusudancs — APPROVED** [review]

**Comment by madhusudancs** [comment]
@irfanurrehman bazel and verification tests are failing. Please take a look.

**Comment by irfanurrehman** [comment]
@madhusudancs updated bazel build files. All tests seem to be passing now. Need lgtm label.

**Comment by madhusudancs** [comment]
@irfanurrehman cool! Added the label.
